pkg/stack/unwind/maps.go

```diff
@@ -0,0 +1,129 @@
+//
+
+package unwind
+
+import (
+	"fmt"
+
+	"github.com/prometheus/procfs"
+)
+
+// ExecutableMapping represents an executable memory mapping.
+type ExecutableMapping struct {
+	LoadAddr   uint64
+	StartAddr  uint64
+	EndAddr    uint64
+	Executable string
+	mainExec   bool
+}
+
+// MainObject returns whether this executable is the "main executable".
+// which triggered the loading of all the other mappings.
+//
+// We care about this because if Linux ASLR is enabled, we have to
+// modify the loaded addresses for the main object.
+func (pm *ExecutableMapping) MainObject() bool {
+	return pm.mainExec
+}
+
+// IsJitted returns whether an executable mapping is JITed or not.
+// The detection is done by checking if the executable mapping is
+// not backed by a file.
+//
+// We don't check for the writeable flag as `mprotect(2)` may be
+// called to make it r+w only.
+func (pm *ExecutableMapping) IsJitted() bool {
+	return pm.Executable == ""
+}
+
+// IsNotFileBacked returns whether the mapping is not backed by a
+// file, such as JIT or vDSO sections.
+func (pm *ExecutableMapping) IsNotFileBacked() bool {
+	return pm.IsJitted() || pm.IsSpecial()
+}
+
+// IsSpecial returns whether the file mapping is a "special" region,
+// such as the mappings for vDSOs `[vdso]` and others.
+func (pm *ExecutableMapping) IsSpecial() bool {
+	return len(pm.Executable) > 0 && pm.Executable[0] == '['
+}
+
+func (pm *ExecutableMapping) String() string {
+	return fmt.Sprintf("ExecutableMapping {LoadAddr: 0x%x, StartAddr: 0x%x, EndAddr: 0x%x, Executable:%s}", pm.LoadAddr, pm.StartAddr, pm.EndAddr, pm.Executable)
+}
+
+type ExecutableMappings []*ExecutableMapping
+
+// HasJitted returns if there's at least one JIT'ed mapping.
+func (pm ExecutableMappings) HasJitted() bool {
+	for _, execMapping := range pm {
+		if execMapping.IsJitted() {
+			return true
+		}
+	}
+	return false
+}
+
+// executableMappingCount returns the number of executable mappings
+// in the passed `rawMappings`.
+func executableMappingCount(rawMappings []*procfs.ProcMap) uint {
+	var executableMappingCount uint
+	for _, rawMapping := range rawMappings {
+		if rawMapping.Perms.Execute {
+			executableMappingCount += 1
+		}
+	}
+	return executableMappingCount
+}
+
+// ExecutableMappings returns the executable memory mappings with the appropriate
+// loaded base address set for non-JIT code.
+//
+// The reason why we need to find the loaded base address is that ELF executables
+// aren't typically loaded in one large executable section, but split in several
+// mappings. For example, the .rodata section, as well as .eh_frame might go in
+// sections without executable permissions, as they aren't needed.
+func ListExecutableMappings(rawMappings []*procfs.ProcMap) ExecutableMappings {
+	result := make([]*ExecutableMapping, 0, executableMappingCount(rawMappings))
+	firstSeen := false
+	for idx, rawMapping := range rawMappings {
+		if rawMapping.Perms.Execute {
+			var LoadAddr uint64
```

Comment:
```suggestion
			var loadAddr uint64
```
I don't know why the linter did not pick up these types of nits.

Reply:
Not sure but perhaps because it's in this inner scope and it being public or protected won't change its visibility?

Reply:
It's a local variable, yes. It's not very common to see an exported local variable. Of course, it won't change anything functionally, but it'd be nice to abide by the convention. Probably, we miss a linter configuration.